js/modules/k6/execution/execution.go

```diff
@@ -40,25 +40,25 @@ type (
 
 	// ModuleInstance represents an instance of the execution module.
 	ModuleInstance struct {
-		modules.InstanceCore
+		modules.InstanceParams
 		obj *goja.Object
 	}
 )
 
 var (
-	_ modules.IsModuleV2 = &RootModule{}
-	_ modules.Instance   = &ModuleInstance{}
+	_ modules.Module   = &RootModule{}
+	_ modules.Instance = &ModuleInstance{}
 )
 
 // New returns a pointer to a new RootModule instance.
 func New() *RootModule {
 	return &RootModule{}
 }
```

Comment:
What I'm going to say would probably be involved and may not be an easy task. And, I'm not deeply knowledgeable about the system and our goals, so this can be a fruitless suggestion, and it might not even work 🤷 

Why don't we register in an extension without giving the k6 an instance of the module, like this:

```
func init() {
	k6modules.Register("k6/x/browser")
}
```

With this, we won't have to stutter in [tests](https://github.com/grafana/k6/blob/092acbbeee509ca9a9b105e89b5e6dddff892723/js/modules/k6/metrics/metrics_test.go#L184-L210), and we can more natural type something like this:

```golang
New(mii)
```

Instead Of:

```golang
New().NewModuleInstance(mii)
```

So the interface would look like this:

```golang
type IsModuleV2 interface {
    New(InstanceCore) Instance
}
```

In an extension we can say:

```golang
type RootModule struct {
    modules.InstanceCore
}

var _ modules.IsModuleV2 = &RootModule{}

func New(m modules.InstanceCore) *RootModule {
	return &RootModule{InstanceCore: m}
}
```

Instead of:

```golang
type (
	RootModule struct{}
	ModuleInstance struct {
		modules.InstanceCore
	}
)

var (
	_ modules.IsModuleV2 = &RootModule{}
	_ modules.Instance   = &ModuleInstance{}
)

func (*RootModule) NewModuleInstance(m modules.InstanceCore) modules.Instance {
	return &ModuleInstance{InstanceCore: m}
}

func New() *RootModule {
	return &RootModule{}
}
```

Of course, with this, we need to create a struct (_RootModule_) of a module ourselves. But then again, we can standardize it like _RootModule_ and instantiate it using reflection.

Reply:
The biggest (and most obvious problem) is that this all si also used for extensions. And while for internal we can have it listed the way it currently works (and likely will still need to work) is that extensions register themselves in a init function and add the `RootModule` so that we can call the `NewModuleInstance` from it ... in reality we can just register the `New` function, but for historical reasons previous to this interface (which is not used by any non-internal extension) you just registered `interface{}` and then k6 did a bunch of reflect "magic"  see #1802, which had a list of problems, the major one of which is ... nobody could understand how it works or why it doesn't work for something or ... why it was needed in some cases.

Hope this explains the problem enough, but I would recommend reading the provided issue.